import pytest

from plenum.common.exceptions import InvalidClientRequest
from plenum.server.plugin.fees.src.client_authnr import FeesAuthNr

#Constants
from plenum.server.plugin.fees.src.constants import SET_FEES
from plenum.server.plugin.token.src.client_authnr import TokenAuthNr
from state.state import State

VALID_ADDR_1 = '6baBEYA94sAphWBA5efEsaA6X2wCdyaH7PXuBtv2H5S1'
VALID_ADDR_2 = '8kjqqnF3m6agp9auU7k4TWAhuGygFAgPzbNH3shp4HFL'
VALID_ADDR_3 = '2LWr7i8bsnE3BgX7MNUNEBpiJYjn7uEA1HNyMSJyy59z'
VALID_ADDR_4 = 'Dy7AxfksXZELTSxF1sUnEocJAJp98E6taMZ9hhLeDtZH'
VALID_ADDR_5 = 'CDe3vTMgns7nXGFwbGkrz24m2VNghmBScdF3QRstF62g'
VALID_ADDR_6 = '62tw4B64yEu7HmJyjMgZ2tE5wgyTtsXTpB7n8VAtfmjL'
VALID_ADDR_7 = 'D2neNhD3mTgN28HKH8qmLBYnaadb9F2GtL5nAzQpkpR6'

VALID_IDENTIFIER = "6ouriXMZkLeHsuXrN1X1fd"
VALID_REQID = 1517423828260117
CONS_TIME = 1518541344
PROTOCOL_VERSION = 1
SIGNATURES = {'B8fV7naUqLATYocqu7yZ8W':
                  '27BVCWvThxMV9pzqz3JepMLVKww7MmreweYjh15LkwvAH4qwYAMbZWeYr6E6LcQexYAikTHo212U1NKtG8Gr2PPP',
              'M9BJDuS24bqbJNvBRsoGg3':
                  '5BzS7J7uSuUePRzLdF5BL5LPvnXxzQyB5BqMT19Hz8QjEyb41Mum71TeNvPW9pKbhnDK12Pciqw9WRHUvsfwdYT5',
              'E7QRhdcnhAwA6E46k9EtZo':
                  'MsZsG2uQHFqMvAsQsx5dnQiqBjvxYS1QsVjqHkbvdS2jPdZQhJfackLQbxQ4RDNUrDBy8Na6yZcKbjK2feun7fg',
              'CA4bVFDU4GLbX8xZju811o':
                  '3A1Pmkox4SzYRavTj9toJtGBr1Jy9JvTTnHz5gkS5dGnY3PhDcsKpQCBfLhYbKqFvpZKaLPGT48LZKzUVY4u78Ki'}


def test_get_fee_idrs_success():
    pass

def test_get_fee_idrs_invalid_data():
    pass

#TODO Needs work to fix this
def test_authenticate_success():
    state = State()
    token_authnr = TokenAuthNr()
    fees_authenticator = FeesAuthNr(state, token_authnr )
    req_data = {'signatures': SIGNATURES, 'reqId': VALID_REQID,
                'operation': {'type': SET_FEES,
                              'fees': {'1': 4, '10001': 8}
                              }
                }
    identifier = VALID_IDENTIFIER

    value = fees_authenticator.authenticate(req_data, VALID_IDENTIFIER)
    assert value == True

def test_authenticate_invalid():
    token_authnr = TokenAuthNr()
    fees_authenticator = FeesAuthNr(None, None)
    req_data = {'signatures': SIGNATURES, 'reqId': VALID_REQID,
                'operation': {'type': 'INVALID_TXN_TYPE',
                              'fees': {'1': 4, '10001': 8}
                              }
                }
    with pytest.raises(InvalidClientRequest):
        fees_authenticator.authenticate(req_data, VALID_IDENTIFIER)

def test_verify_signature_success():
    pass

def test_verify_signature_invalid():
    pass
